Production/src/Selectors.cc

```diff
@@ -177,5 +179,94 @@ TauJetSelector::Result TauJetTag::Select(const edm::Event& event, const std::deq
+        if(!(muon.pt() > 28 && std::abs(muon.eta()) < 2.1 && muon.isMediumMuon() && PFRelIsolation(muon) < 0.15
+                && std::abs(muon.muonBestTrack()->dz(primaryVertex.position())) < 0.2
+                && std::abs(muon.muonBestTrack()->dxy(primaryVertex.position())) < 0.045))
+            continue;
+        if(!ref_muon || PFRelIsolation(*ref_muon) > PFRelIsolation(muon) || (PFRelIsolation(*ref_muon) == PFRelIsolation(muon) && ref_muon->pt() < muon.pt())){
+            ref_muon = &muon;
+        }
+    }
+    if(!ref_muon) {
+    	for(const pat::Electron& electron : electrons) { // Start looking for an ele tag if no muon was found
+            if(!(electron.pt() > 34 && std::abs(electron.eta()) < 2.3 && electron.electronID("mvaEleID-Fall17-noIso-V2-wp90") > 0.5f && PFRelIsolation(electron,rho) < 0.15
+                     && std::abs(electron.bestTrack()->dz(primaryVertex.position())) < 0.2
+                     && std::abs(electron.bestTrack()->dxy(primaryVertex.position())) < 0.045))
+            	 continue;
+            if(!ref_electron || PFRelIsolation(*ref_electron,rho) > PFRelIsolation(electron,rho) || (PFRelIsolation(*ref_electron,rho) == PFRelIsolation(electron,rho) && ref_electron->pt() < electron.pt())){
+                ref_electron = &electron;
+            }
+        }
+    }
+    //
+    auto tagObject = std::make_shared<TagObject>();
+    std::vector<const TauJet*> selectedTauJets;
+    if(ref_muon || ref_electron || met.pt() > 180) {
+    	if(ref_muon) { // Fill tag object with muon if muon found
+        	tagObject->kind = tau_analysis::selectors::TagObject::Kind::Muon; //new tag object member 
+        	tagObject->p4 = ref_muon->polarP4();
+       		tagObject->charge = ref_muon->charge();
+        	tagObject->id = unsigned(ref_muon->isLooseMuon()) * 1 + unsigned(ref_muon->isMediumMuon()) * 2
+                          	+ unsigned(ref_muon->isTightMuon(primaryVertex)) * 4;
+        	tagObject->isolation = PFRelIsolation(*ref_muon);
+        	tagObject->has_extramuon = hasExtraMuon(muons, ref_muon, primaryVertex);
+        	tagObject->has_extraelectron = hasExtraElectron(electrons, rho);
+        	tagObject->has_dimuon = hasExtraDimuon(muons, primaryVertex);
+    	}
+    	//	
+    	else if(ref_electron) { // Fill tag with ele if ele found rather than muon
+		tagObject->kind = tau_analysis::selectors::TagObject::Kind::Electron; //new tag object member
+        	tagObject->p4 = ref_electron->polarP4();
+        	tagObject->charge = ref_electron->charge();
+        	tagObject->id = ref_electron->electronID("mvaEleID-Fall17-noIso-V2-wp90"); //Not sure what to put here 
+        	tagObject->isolation = PFRelIsolation(*ref_electron,rho);
+        	tagObject->has_extramuon = hasExtraMuon(muons, ref_muon, primaryVertex);
+        	tagObject->has_extraelectron = hasExtraElectron(electrons, rho);
+        	tagObject->has_dimuon = hasExtraDimuon(muons, primaryVertex);	
+    	}
+    	else { // high MET case
```

Comment:
It would be nice to have some flag "which of the selection group it is" (1), (2), (3) or (4) since now tagObject kind is only valid for muon/electron tag

Reply:
I see, but in this case there might be confusion, since: tag Object is filled even though MET or gen lepton might not be a tag object in general sense